physicsnemo/optim/combined_optimizer.py

```diff
@@ -0,0 +1,284 @@
+    * **Dynamic Parameter Addition**: The ``add_param_group()`` method is not
+      supported. To add parameters dynamically, add them to the individual
+      optimizers before creating the CombinedOptimizer, or create a new instance.
+    * **State Access**: The ``state`` attribute inherited from the base class may
+      not accurately reflect the optimizer state. Access state through the
+      individual optimizers in the ``optimizers`` attribute instead.
+    * **Serialization**: The optimizer can be pickled and unpickled. When
+      ``torch_compile_kwargs`` is provided, the compiled step functions are
+      reconstructed during unpickling.
+
+    Examples
+    --------
+    Combine Adam for model backbone and SGD for the head:
+
+    >>> import torch
+    >>> import torch.nn as nn
+    >>> from torch.optim import Adam, SGD
+    >>> from physicsnemo.optim import CombinedOptimizer
+    >>>
+    >>> model = nn.Sequential(
+    ...     nn.Linear(10, 20),  # backbone
+    ...     nn.ReLU(),
+    ...     nn.Linear(20, 2),   # head
+    ... )
+    >>> backbone_params = list(model[0].parameters())
+    >>> head_params = list(model[2].parameters())
+    >>>
+    >>> opt1 = Adam(backbone_params, lr=1e-4)
+    >>> opt2 = SGD(head_params, lr=1e-2, momentum=0.9)
+    >>> combined_opt = CombinedOptimizer([opt1, opt2])
+    >>>
+    >>> # Use with a learning rate scheduler
+    >>> scheduler = torch.optim.lr_scheduler.StepLR(combined_opt, step_size=10)
+    >>>
+    >>> # Standard training loop
+    >>> for epoch in range(100):
+    ...     combined_opt.zero_grad()
+    ...     loss = model(torch.randn(32, 10)).sum()
+    ...     loss.backward()
+    ...     combined_opt.step()
+    ...     scheduler.step()
+    """
+
+    def __init__(
+        self,
+        optimizers: Sequence[Optimizer],
+        torch_compile_kwargs: dict[str, Any] | None = None,
+    ):
+        if not optimizers:
+            raise ValueError("`optimizers` must contain at least one optimizer.")
+
+        self.optimizers = optimizers
+        self._torch_compile_kwargs = torch_compile_kwargs
+
+        ### Aggregate parameter groups from all optimizers
+        # We pass an empty defaults dict because hyperparameters are managed by
+        # the individual optimizers, not this wrapper.
+        param_groups = [g for opt in optimizers for g in opt.param_groups]
+
+        # Flag to allow add_param_group during initialization
+        self._initializing = True
+        try:
+            super().__init__(param_groups, defaults={})
+        finally:
+            self._initializing = False
+
+        ### Setup step functions (optionally compiled)
+        if torch_compile_kwargs is None:
+            self.step_fns: list[Callable] = [opt.step for opt in optimizers]
+        else:
+            self.step_fns: list[Callable] = [
+                torch.compile(opt.step, **torch_compile_kwargs) for opt in optimizers
+            ]
+
+    def zero_grad(self, *args, **kwargs) -> None:
```

Comment:
Upstream interface accepts only `set_to_none=True` here, and not other parameters.  I know if we passed it, that would succeed, but unless there are optimizers accepting other vales for `zero_grad` I think we should stick to `set_to_none`.

Reply:
Fixed 7ddf8ab4df5b192a8ea694d45529bd9b5549f918